tap/api/api.go

```diff
@@ -0,0 +1,161 @@
+package api
+
+import (
+	"bufio"
+	"plugin"
+	"time"
+)
+
+type Protocol struct {
+	Name            string   `json:"name"`
+	LongName        string   `json:"long_name"`
+	Abbreviation    string   `json:"abbreviation"`
+	Version         string   `json:"version"`
+	BackgroundColor string   `json:"background_color"`
+	ForegroundColor string   `json:"foreground_color"`
+	FontSize        int8     `json:"font_size"`
+	ReferenceLink   string   `json:"reference_link"`
+	Ports           []string `json:"ports"`
+	Priority        uint8    `json:"priority"`
+}
+
+type Extension struct {
+	Protocol  Protocol
+	Path      string
+	Plug      *plugin.Plugin
+	Dissector Dissector
+}
+
+type ConnectionInfo struct {
+	ClientIP   string
+	ClientPort string
+	ServerIP   string
+	ServerPort string
+	IsOutgoing bool
+}
+
+type TcpID struct {
+	SrcIP   string
+	DstIP   string
+	SrcPort string
+	DstPort string
+	Ident   string
+}
+
+type CounterPair struct {
+	Request  uint
+	Response uint
+}
+
+type GenericMessage struct {
+	IsRequest   bool        `json:"is_request"`
+	CaptureTime time.Time   `json:"capture_time"`
+	Payload     interface{} `json:"payload"`
+}
+
+type RequestResponsePair struct {
+	Request  GenericMessage `json:"request"`
+	Response GenericMessage `json:"response"`
+}
+
+type OutputChannelItem struct {
+	Protocol       Protocol
+	Timestamp      int64
+	ConnectionInfo *ConnectionInfo
+	Pair           *RequestResponsePair
```

Comment:
@nimrod-up9 not passing a pointer there was intentional. The `Protocol` is initially assigned by the loaded extensions but then further modified in the later stages of data propagation. I didn't wanted to expose the memory region of extension properties to other parts of the codebase. It can be turned into `*Protocol` though, it would improve the performance a little bit. But I'm not sure about the necessary changes at this moment.

Reply:
Fair point. Can you add that in a comment why Protocol should not be a pointer?

Reply:
@nimrod-up9 added a comment with https://github.com/up9inc/mizu/commit/e27faac6d46ca9237a32f7df36e50e43c92586e4